### PR TITLE
Switch epoll and kqueue to level triggers

### LINKED ISSUE #2654: Poll based datapaths can hang sends in race condition with kernel
### Describe the bug

For sends in the datapath, we don't enable the edge triggers until after a socket write returns EAGAIN or EWOULDBLOCK. However, if sends in the kernel become enabled between calling send() and epoll_ctl (Or equivalent in other paths), the epoll event would never tirgger, and the sends would be hung forever for that socket.

### Affected OS

- [ ] All
- [ ] Windows Server 2022
- [ ] Windows 11
- [ ] Windows Insider Preview (specify affected build below)
- [X] Ubuntu
- [ ] Debian
- [X] Other (specify below)

### Additional OS information

kqueue and epoll datapaths, in addition to PSN on windows

### MsQuic version

main

### Steps taken to reproduce bug

Race condition

### Expected behavior

Sends can't hang

### Actual outcome

Sends can hang

### Additional details

The solution to this will be to always keep the edge triggers enabled. Then before we send, enqueue the send into the pending queue. If the send passes, dequeue it, otherwise just continue, and the datapath thread will pick up the send later.

### PR DESCRIPTION
**Pull request by thhous-msft**
## Description

kqueue and epoll datapaths have a race condition upon pended sends whereas they might not get trigger notifications. By switching to level triggers, we solve this issue.

Closes #2654

## Testing

Existing tests cover the change.

## Documentation

N/A

### CODE REVIEW

**Comment by thhous-msft** [comment]
Don't merge this yet, I want to do more local testing. Its just easy for CI to test everything as well.

**Review by anrossi — DISMISSED** [review]

**Comment by nibanks** [comment]
I still have my reservations about making changes here, because we've never seen an issue with the current code. Is there any test we can write (maybe at the datapath unit test layer) to prove/disprove this change?

**Comment by thhous-msft** [comment]
Well we validate that it works, and doesn't break performance. If it was going to be problematic, it would have triggered an issue with those cases. And using a TCP test, validated that the logic holds.

The issue is its basically impossible to trigger EAGAIN on write for UDP. I've tried looking if theres a way to force it, but I can't find any.

**Review by nibanks — DISMISSED** [review]

**Comment by nibanks** [comment]
Ok. I leave you in charge of the final decision!

**Comment by thhous-msft** [comment]
The discussion created in https://github.com/microsoft/msquic/discussions/2662 is another reason for doing this change. Level triggered mode is required to make it so the entire receive buffer does not have to be read every time an event is received.

**Review by nibanks — APPROVED** [review]

**Review by anrossi — APPROVED** [review]

**Review by nibanks — APPROVED** [review]

**Pull request merged**
